Eventonica (#4)
priyaraj7/curriculum: project-example/week-9/eventonica/eventonica-express-backend.md

### Instructions

#### Create a new Express App

1. In your terminal, navigate to your `eventonica/client` directory . Start it with `npm start`.

1. You will need to create a second app for your Express backend. In a second terminal window, navigate to your general `Eventonica` project folder. Follow these commands to create a new project called `server` and start it.

   ```
   mkdir server
   cd server
   npm init -y
   ```

1. Let’s install Express, nodemon and cors as a dependency to use it

   ```shell

   npm install express cors
   npm install --save-dev nodemon
   ```

- Let’s take a quick look at the three packages:

  - express: Express is a fast and lightweight web framework for Node.js. Express is an essential part of the PERN stack.

>   - cors: CORS is a node.js package for providing an Express middleware that can be used to enable CORS with various options. Cross-origin resource sharing (CORS) is a mechanism that allows restricted resources on a web page to be requested from another domain outside the domain from which the first resource was served.

Reviewer: Nice addition! It could also use an explanation of why we need it in our case. E.g. We need cors because we're running the client and server apps on different domains (because the ports are different). If they were running on the same domain, we wouldn't need to use this.

  - nodemon: nodemon is a tool that helps develop Node.js based applications by automatically restarting the node application when file changes in the directory are detected.

> - Next, go into the package.json `scripts` and add this start script. This help you to run server with a simple command. Tell Node.js that all files are ES Modules by adding `"type": "module"`

Reviewer: This section starts out as a numbered list, but then switches to a bullet list. Probably should switch this back to a number:
```suggestion
1. Next, go into the package.json `scripts` and add this start script. This helps you to run the server with a simple command. Tell Node.js that all files are ES Modules by adding `"type": "module"`
```

1. Testing the endpoint
   Open Postman, and open a new tab called "Eventonica". "GET" should be automatically selected in the dropdown. Enter `http://localhost:4000/api/users` into the request URL and press send. You should see the same list of users in the response below.

2. Testing sending information
   Now the get users endpoint always returns all users. However, in the future we might want filtering.

To test this, you can console log `console.log(req.body, 'the body')` before the `res.json` line. In Postman, now try adding a body. Click the "body" tab and select "raw". Then select "JSON" from the dropdown. Try sending a JSON of something that the API might send. For example, it could send `{"name": "nemo"}` for when you implement filtering.
Right now you can see `undefined` in console. Add Express body parser `app.use(express.json());` in index.js. Now you can see `{"name": "nemo"}` in console.
Note: that console logs will not show up in Postman.

> Your final code look something like this:

Reviewer: ```suggestion
Your final code should look something like this:
```

    import React, { useEffect, useState } from 'react';
    ```

1.  If you visit localhost:3000, you should see the response from your `/api/users` route!

#### Use your API data to render a events list in your React app

Now your challenge is to:

- Move your example events out of `client/src/components/Events.jsx` and into `server/index.js` and make sure it is a single array of events.
- Have the events array be the response from http://localhost:4000/api/events, and make sure it renders in your frontend on localhost:3000
- Have your React Events component render events as HTML list items rather than plain text.

#### The real work

Add remaining REST API routes for `users` listed in the [project README](./README.md). Don't worry about `events` or `favorites` for now... you can add those after we get our database going.

`Users` now needs "add" and "delete" functionality. For example, a frontend function called `addUser()` should make a POST request to http://localhost:4000/api/users/
and add a user by posting JSON with your API (which currently just saves to a variable since we have no DB), and the API would need a route like this:

```js

> app.post('/', function (req, res, next) {

Reviewer: ```suggestion
app.post('/api/users', function (req, res, next) {
```
